### ISSUE TITLE
Build snippets failed for `ResourceManager` and `Azure.Core`

### ISSUE THREAD

**Issue by fengzhou-msft** [Azure/azure-sdk-for-net#33945]
See [CI error](https://dev.azure.com/azure-sdk/public/_build/results?buildId=2168366&view=logs&j=b70e5e73-bbb6-5567-0939-8415943fadb9&t=9069f0a2-bd31-51b1-8b5c-08355cf80229&l=118).

> sdk/resourcemanager/Azure.ResourceManager/tests/Samples/Readme.cs(26,9): error CS1022: Type or namespace definition, or end-of-file expected 
> sdk/resourcemanager/Azure.ResourceManager/tests/Samples/Readme.cs(30,9): error CS0106: The modifier 'public' is not valid for this item
> sdk/core/Azure.Core/tests/samples/PollyPolicy.cs(75,1): error CS1022: Type or namespace definition, or end-of-file expected

It's probably related to the use of `#if SNIPPET` or `#if !SNIPPET` in following files:
[Readme.cs](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/resourcemanager/Azure.ResourceManager/tests/Samples/Readme.cs)
[PollyPolicy.cs](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/tests/samples/PollyPolicy.cs)

**Comment by fengzhou-msft** [comment]
@heaths can you advise on how to fix the issue?

**Comment by jsquire** [comment]
//cc: @JoshLove-msft


The Polly Policy code appears to be invalid when `SNIPPET` is defined.  I believe it needs an `#else` that hides the rest of the code.    Likewise, for `ReadMe.cs` - when `SNIPPET` is defined, the code is malformed and not compilable.

**Comment by JoshLove-msft** [comment]
Hmm I'm not sure how the change to always build snippets was able to merge in with build failures.

**Comment by heaths** [comment]
@JoshLove-msft I had the same problem with Cognitive. I'm assuming it's because we changed how `net - core - ci` works now, that it doesn't necessarily build everything.

@fengzhou-msft the code - whether built with `SNIPPET` or not - has to be valid code. There's lots of tricks you can do, but generally think about either defining things uniquely in any `#if SNIPPET` and `#else` block, or define them generally and override in an `#if !SNIPPET` block. Use the command line from the pipeline failure e.g., `dotnet build eng/service.proj ...` to test your changes before submitting to avoid pushing commit after commit until the CI works (faster, too).

**Comment by JoshLove-msft** [comment]
net - core - ci doesn't build everything? I thought we split the tests out among different pipelines but that we would still build everything.
/cc @weshaggard @m-nash

**Comment by weshaggard** [comment]
@JoshLove-msft snippets are only built in the Analyze step not during the normal source or test build which is why it wasn't covered by the core ci.

**Comment by heaths** [comment]
It also didn't occur because sdk/core/ci.yml had `BuildSnippets: false` because of some of its own errors that didn't have a straight forward solution. Either way, these were missed. I'm working on fixing a few and will turn off snippet build for `*` as well as individual project directories if not immediately easy to fix.

**Comment by m-nash** [comment]
@weshaggard we might consider doing the same thing we did with resourcemanager where the leaf projects run but only the build and analyze steps which are much faster.

**Comment by weshaggard** [comment]
That wouldn't have helped here as those other pipelines weren't triggered because it was a common file that changed and not anything in their directory.

**Comment by m-nash** [comment]
Yeah my suggestion was around updating triggers so that they do run off common files :-)

Or maybe the solution is to build a 2nd sparse matrix that handles build / analyze while the current one handles tests.
